### ISSUE TITLE
Can we Save player's state during ddos, And restore after ddos end

### ISSUE THREAD

**Issue by QingGo** [ddnet/ddnet#3780]
Now after the ddnet server is hit into the black hole by ddos, the players who are playing cannot timeout back after the black hole ends, because the black hole time is too long.
I have an idea, during the black hole, can I send SIGTSTP to the ddnet-server process (just like ctrl + z), and then use the fg command to recover after the black hole ends.

I also thought about dumping the memory and immediately recovering it on another machine, but this is probably more difficult.
Or dump the necessary information, and then load it when another machine starts, but this has to change the code.
So SIGTSTP is probably relatively easy to implement, just need a background script.

But after testing, using the fg command after ctrl + z for a period of time will trigger timed out and Timeout Protection over at the same time, so we have to change the code anyway to achieve this effect.
![image](https://user-images.githubusercontent.com/13678719/115121808-3f1c1980-9fe7-11eb-9a68-64c39ebc9a90.png)

**Comment by def-** [comment]
we could extend timeout protection automatically when a dos is happening
Can be done externally and by just writing into the fifo files. How to best monitor that an attack is happening? Something like > X packets / second incoming?

**Comment by QingGo** [comment]
> we could extend timeout protection automatically when a dos is happening
> Can be done externally and by just writing into the fifo files. How to best monitor that an attack is happening? Something like > X packets / second incoming?

I think maybe X packets / second incoming is not a very good metric. because the virtual machine bandwidth is limited by cloud service provider, the number of packets received by the virtual machine is limited either.
Maybe we can detect whether the server can successfully send the data packet and receive the response, for example, check if ddnet server can register to the master servert.

**Comment by def-** [comment]
Often the DDNet server will think it's connected to master server, or only specific packets are being blocked by the ddos protection, so I'm not sure if this works much better. The master servers also get attacked more often.

Maybe we can try pinging something and checking the packets / second, and when either of them is bad, increase the timeout protection for the next X minutes.

**Comment by QingGo** [comment]
> Often the DDNet server will think it's connected to master server, or only specific packets are being blocked by the ddos protection, so I'm not sure if this works much better. The master servers also get attacked more often.
> 
> Maybe we can try pinging something and checking the packets / second, and when either of them is bad, increase the timeout protection for the next X minutes.

nice idea. ping something, then check the packet loss rate during a period of time(such as 20 seconds), if packet loss rate higher than some limit(such as 50%), then increase the timeout protection.

**Comment by QingGo** [comment]
Also, when the server status returns to normal, maybe we should also keep the timeout protection  a period of time(maybe 12 hours?) so that players can know about the server recovery information and re-enter the game.

**Comment by def-** [comment]
12 hours is a bit excessive to be honest. We would end up with servers being quite full and you could even consider this as faking a higher player number. (Maybe we should not count timed out players to the player count.)

**Comment by QingGo** [comment]
> 12 hours is a bit excessive to be honest. We would end up with servers being quite full and you could even consider this as faking a higher player number. (Maybe we should not count timed out players to the player count.)

Actually, I think it’s a good idea to not counting timeout players  as  the actual number of players. Sometimes I timeout on a 16-player server, because the player is full, and I can’t re-enter the server to recover. 
However, maybe it would be difficult to change that part of the code?

**Comment by def-** [comment]
Even if we don't count them, you couldn't reenter since they still take up a slot on the server.

**Comment by QingGo** [comment]
> Even if we don't count them, you couldn't reenter since they still take up a slot on the server.

yes. timeout player will take up a slot on the server, that will be an another issue(and maybe hard to fix). At now i think extending the protection time is already a big progress. (if 12 hours is too long, I think 3 hours is also fine.)

**Comment by def-** [comment]
I think I'm going to try something like this on the servers:
```python
#!/usr/bin/env python3
import os
import time
import collections

INTERVAL = 10

# Traffic class adapted from BotoX' ServerStatus: https://github.com/BotoX/ServerStatus
class Traffic:
    def __init__(self):
        self.rx = collections.deque(maxlen=10)
    def get(self):
        f = open('/proc/net/dev', 'r')
        net_dev = f.readlines()
        f.close()
        avgrx = 0

        for dev in net_dev[2:]:
            dev = dev.split(':')
            if dev[0].strip() == "lo" or dev[0].find("tun") > -1:
                continue
            dev = dev[1].split()
            avgrx += int(dev[0])

        self.rx.append(avgrx)
        avgrx = 0

        l = len(self.rx)
        for x in range(l - 1):
            avgrx += self.rx[x+1] - self.rx[x]

        avgrx = int(avgrx / l / INTERVAL)

        return avgrx

def main():
    traffic = Traffic()
    traffic.get()
    bad = 0
    while True:
        netrx = traffic.get()
        ping = os.system("ping -c 1 master.ddnet.tw > /dev/null") == 0
        if netrx > 3_000_000 or not ping:
            bad += 1
        if bad == 0:
            os.system("for i in /home/teeworlds/servers/servers/*.fifo; do echo conn_timeout_protection 3600 > $i & done")
            bad = 0
        print(bad)
        time.sleep(INTERVAL)

if __name__ == '__main__':
    main()
```
If network is bad for a while, timeout protection is increased from 16 min to 1 hour. We don't set it back down, but servers are restarted once a day. I'll monitor how often it happens on the servers and adapt it based on that.

**Comment by QingGo** [comment]
> I think I'm going to try something like this on the servers:
> 
> ```python
> #!/usr/bin/env python3
> import os
> import time
> import collections
> 
> INTERVAL = 10
> 
> # Traffic class adapted from BotoX' ServerStatus: https://github.com/BotoX/ServerStatus
> class Traffic:
>     def __init__(self):
>         self.rx = collections.deque(maxlen=10)
>     def get(self):
>         f = open('/proc/net/dev', 'r')
>         net_dev = f.readlines()
>         f.close()
>         avgrx = 0
> 
>         for dev in net_dev[2:]:
>             dev = dev.split(':')
>             if dev[0].strip() == "lo" or dev[0].find("tun") > -1:
>                 continue
>             dev = dev[1].split()
>             avgrx += int(dev[0])
> 
>         self.rx.append(avgrx)
>         avgrx = 0
> 
>         l = len(self.rx)
>         for x in range(l - 1):
>             avgrx += self.rx[x+1] - self.rx[x]
> 
>         avgrx = int(avgrx / l / INTERVAL)
> 
>         return avgrx
> 
> def main():
>     traffic = Traffic()
>     traffic.get()
>     bad = 0
>     while True:
>         netrx = traffic.get()
>         ping = os.system("ping -c 1 master.ddnet.tw > /dev/null") == 0
>         if netrx > 3_000_000 or not ping:
>             bad += 1
>         if bad > 10:
>             os.system("for i in /home/teeworlds/servers/servers/*.fifo; do echo conn_timeout_protection 3600 > $i & done")
>             bad = 0
>         time.sleep(INTERVAL)
> 
> if __name__ == '__main__':
>     main()
> ```
> 
> If network is bad for a while, timeout protection is increased from 16 min to 1 hour. We don't set it back down, but servers are restarted once a day. I'll monitor how often it happens on the servers and adapt it based on that.

good job. thanks!

**Comment by QingGo** [comment]
> I think I'm going to try something like this on the servers:
> 
> ```python
> #!/usr/bin/env python3
> import os
> import time
> import collections
> 
> INTERVAL = 10
> 
> # Traffic class adapted from BotoX' ServerStatus: https://github.com/BotoX/ServerStatus
> class Traffic:
>     def __init__(self):
>         self.rx = collections.deque(maxlen=10)
>     def get(self):
>         f = open('/proc/net/dev', 'r')
>         net_dev = f.readlines()
>         f.close()
>         avgrx = 0
> 
>         for dev in net_dev[2:]:
>             dev = dev.split(':')
>             if dev[0].strip() == "lo" or dev[0].find("tun") > -1:
>                 continue
>             dev = dev[1].split()
>             avgrx += int(dev[0])
> 
>         self.rx.append(avgrx)
>         avgrx = 0
> 
>         l = len(self.rx)
>         for x in range(l - 1):
>             avgrx += self.rx[x+1] - self.rx[x]
> 
>         avgrx = int(avgrx / l / INTERVAL)
> 
>         return avgrx
> 
> def main():
>     traffic = Traffic()
>     traffic.get()
>     bad = 0
>     while True:
>         netrx = traffic.get()
>         ping = os.system("ping -c 1 master.ddnet.tw > /dev/null") == 0
>         if netrx > 3_000_000 or not ping:
>             bad += 1
>         if bad > 10:
>             os.system("for i in /home/teeworlds/servers/servers/*.fifo; do echo conn_timeout_protection 3600 > $i & done")
>             bad = 0
>         time.sleep(INTERVAL)
> 
> if __name__ == '__main__':
>     main()
> ```
> 
> If network is bad for a while, timeout protection is increased from 16 min to 1 hour. We don't set it back down, but servers are restarted once a day. I'll monitor how often it happens on the servers and adapt it based on that.

I ask DDNet CHN server hosters and know that when being DDOS the first time, the during of blackhole time will be 5 hours. So maybe 1 hour is not enough, maybe we should keep monitoring the network situation, if it has not recovered, then continue to accumulate the protection time? for example:

``` python3
def main():
    traffic = Traffic()
    traffic.get()
    bad = 0
    timeout_protection_time = 3600
    is_last_state_timeout = True
    while True:
        netrx = traffic.get()
        ping = os.system("ping -c 1 master.ddnet.tw > /dev/null") == 0
        if netrx > 3_000_000 or not ping:
            bad += 1
        if bad > 10:
            if is_last_state_timeout:
                timeout_protection_time += INTERVAL * 2
            os.system("for i in /home/teeworlds/servers/servers/*.fifo; do echo conn_timeout_protection {} > $i & done".format(timeout_protection_time))
            is_last_state_timeout = True
            bad = 0
        else:
            is_last_state_timeout = False
            timeout_protection_time -= INTERVAL
        time.sleep(INTERVAL)
```

**Comment by def-** [comment]
Keeping it alive for > 5 hours will be a waste. The blackholing also often lasts for 12-24 hours, we don't have a fix for that. And if we would save people in team 0 there will be cheats possible.

**Comment by QingGo** [comment]
> Keeping it alive for > 5 hours will be a waste. The blackholing also often lasts for 12-24 hours, we don't have a fix for that. And if we would save people in team 0 there will be cheats possible.


When the server is working normally, 5 hours may be a waste. But in the black hole state, new players cannot enter the game anyway, so timeout players will not occupy the game resources of normal players. Maybe we can think a mechanism that allows the timeout protection to last only one hour after recovering from the black hole state.
Also, if the player uses the timeout mechanism to return to the game. What we have done are just extending the timeout protection time, I can't think that this will increase the possibility of any cheating.
